### PR TITLE
Make XBlock due date and grace period aware

### PR DESCRIPTION
**Pull request by saharm1**
@stvstnfrd @caesar2164 

cc @tkim135

### CODE REVIEW

**Review by caesar2164 — APPROVED** [review]
Looks good to me! 👍

**Comment by saharm1** [comment]
Latest commit fixes this issue: 
screen shot 2018-07-19 at 11 02 38 am

screen shot 2018-07-23 at 2 23 59 pm

**Review by caesar2164 — CHANGES_REQUESTED** [review]
Let's move the edit file as well, since we're at it, @saharm1.

File: submit_and_compare/mixins.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        if due is not None:
+            # Remove timezone information from platform provided due date.
+            # Dates are stored as UTC timezone aware objects on platform.
+            due = due.replace(tzinfo=None)
+            if graceperiod is not None:
+                # Compare the datetime objects (both have to be timezone naive)
+                due = due + graceperiod
+            return now > due
+        return False
+
```
Comment (stvstnfrd) [inline submit_and_compare/mixins.py]: Extra newline?

**Review by caesar2164 — APPROVED** [review]
cool, nice @saharm1!

I'm 👍, you just need @stvstnfrd to approve now.

File: submit_and_compare/submit_and_compare.py [hunk 2]
```diff
@@ -193,6 +196,19 @@ class SubmitAndCompareXBlock(XBlock):
 
     has_score = True
 
+    def build_fragment(
+            self,
+            template, 
+            context_dict,
+    ):
+        #  pylint: disable=dangerous-default-value, too-many-arguments
```
Comment (stvstnfrd) [inline submit_and_compare/submit_and_compare.py]: See comment on the other PR about whether we still need this..

File: submit_and_compare/submit_and_compare.py [hunk 3]
```diff
@@ -313,14 +332,14 @@ def student_submit(self, submissions, suffix=''):
...
+            success = True
 
-            result = {
-                'success': True,
-                'problem_progress': self._get_problem_progress(),
-                'submit_class': self._get_submit_class(),
-                'used_attempts_feedback': self._get_used_attempts_feedback(),
-            }
-        return result
+        return {
```
Comment (stvstnfrd) [inline submit_and_compare/submit_and_compare.py]: There's still benefits to assigning this to a local variable and then returning that.
Code that immediately returns a new collection like this can be a pain to step through interactively with a debugger, should the need arise.

```py
result = { ... }
return result
```

File: submit_and_compare/submit_and_compare.py [hunk 4]
```diff
@@ -289,17 +309,16 @@ def student_submit(self, submissions, suffix=''):
...
                 'number of allowed attempts',
             )
-            result = {
-                'success': False,
-                'problem_progress': self._get_problem_progress(),
-                'submit_class': self._get_submit_class(),
-                'used_attempts_feedback': self._get_used_attempts_feedback(),
-            }
+        elif self.is_past_due():
+            LOG.debug(
```
Comment (stvstnfrd) [inline submit_and_compare/submit_and_compare.py:318]: Debug seems like an appropriate level of logging here :)

File: submit_and_compare/submit_and_compare.py [hunk 5]
```diff
@@ -28,7 +31,7 @@ def get_body(xmlstring):
     """
     tree = etree.parse(StringIO(xmlstring))
     body = tree.xpath('/submit_and_compare/body')
-    body_string = etree.tostring(body[0], encoding='unicode')
+    body_string = etree.tostring(body[0], method='text', encoding='unicode')
```
Comment (stvstnfrd) [inline submit_and_compare/submit_and_compare.py:34]: Why this change?

File: .gitignore [hunk 6]
```diff
@@ -4,3 +4,4 @@
 *.pyc
 *.sw[op]
 coverage.xml
+.DS_STORE
```
Comment (stvstnfrd) [inline .gitignore]: Alternatively, since this is an OSX-specific file and not something inherent to this project, you could add this to your personal ignore file in `~/.gitignore`.
Then, those files will be ignored for any repo you work on.

CC: @tkim135

**Review by stvstnfrd — CHANGES_REQUESTED** [review]

File: submit_and_compare/submit_and_compare.py (hunk 5 above)
Comment (saharm1) [reply submit_and_compare/submit_and_compare.py:34]: @stvstnfrd there was a strange error, see my comment with the pictures above

**Review by stvstnfrd — APPROVED** [review]
